ambry-store/src/main/java/com/github/ambry/store/PersistentIndex.java

```diff
@@ -1643,8 +1643,10 @@ private StoreFindToken findEntriesFromSegmentStartOffset(Offset initialSegmentSt
         break; // we have entered and finished reading from the journal, so we are done.
       }
 
-      if (segmentStartOffset == validIndexSegments.lastKey() && !onlyContainsHeaderInLastLogSegment(
-          segmentStartOffset)) {
+      //if the journal based token point to the last entry of last index segment but journal cleaned up after to last log
+      //segment auto closed, we can bypass this check and return the index based token.
+      if (segmentStartOffset == validIndexSegments.lastKey() && !onlyContainsHeaderInLastLogSegment(segmentStartOffset,
```

Comment:
I think this should be changed to `segmentStartOffet.equals(validIndexSemgents.lastKey())`. Prior to your change, if we reach the last index segment, the `segmentStartOffset` should be extracted by calling `indexSegments.higherKey(segmentStartOffset)` at line 1602, but now, it might just be the offset in storeFindToken we can't just compare the reference any more, we have to compare the actual value.

Reply:
Updated.